packages/sparrowql/src/build.ts

```diff
@@ -30,301 +50,390 @@ export type Options = {
+    const toCheck = relations.filter(relation => toJoin.includes(relation.to));
+    if (toCheck.length === 0) {
+      throw new Error(
+        `Needed connection to ${toJoin
+          .map(name => `"${name}"`)
+          .join(', ')} but no way to connect.`,
+      );
+    }
 
-    let $ = 0;
-    while ($++ < 100) {
-        inlineSort();
+    const { cost, path } = closestPath(relations, toCheck, joined, query);
+    if (isFinite(cost)) {
+      path.forEach(single => {
+        inlineRelation(single);
         inlineMatch();
         inlineGroup();
         inlineSort();
         inlineLimitAndSkip();
-
-        const toJoin = needed.filter(name => !joined.includes(name));
-        if (toJoin.length === 0) break;
-
-        const toCheck = relations.filter(relation => toJoin.includes(relation.to));
-        if (toCheck.length === 0)
-            throw new Error(
-                `Needed connection to ${toJoin.map(name => `"${name}"`).join(', ')} but no way to connect.`
-            );
-
-        const {cost, path} = closestPath(relations, toCheck, joined, query);
-        if (isFinite(cost)) {
-            path.forEach(single => {
-                inlineRelation(single);
-                inlineMatch();
-                inlineGroup();
-                inlineSort();
-                inlineLimitAndSkip();
-            });
-            continue;
-        }
-
-        throw new Error(`Needed connection to ${toJoin.map(name => `"${name}"`).join(', ')} but no way to connect.`);
+      });
+      continue;
     }
 
-    inlineLimitAndSkip();
-
-    // Optimize sort.
-    if (isBaseSort && steps.some(step => 'limit' in step) && steps.some(step => 'sort' in step)) {
-        const indexMatch = steps.findIndex(step => 'match' in step);
-
-        steps.splice(indexMatch + 1, 0, steps[steps.findIndex(step => 'sort' in step)]);
+    throw new Error(
+      `Needed connection to ${toJoin
+        .map(name => `"${name}"`)
+        .join(', ')} but no way to connect.`,
+    );
+  }
+
+  inlineLimitAndSkip();
+
+  // Optimize sort.
+  if (
+    isBaseSort &&
+    steps.some(step => 'limit' in step) &&
+    steps.some(step => 'sort' in step)
+  ) {
+    const indexMatch = steps.findIndex(step => 'match' in step);
+
+    steps.splice(
+      indexMatch + 1,
+      0,
+      steps[steps.findIndex(step => 'sort' in step)],
+    );
+
+    if (steps.some(step => 'skip' in step)) {
+      steps.splice(
+        indexMatch + 2,
+        0,
+        steps.splice(
+          steps.findIndex(step => 'skip' in step),
+          1,
+        )[0],
+      );
+      steps.splice(
+        indexMatch + 3,
+        0,
+        steps.splice(
+          steps.findIndex(step => 'limit' in step),
+          1,
+        )[0],
+      );
+    } else {
+      steps.splice(
+        indexMatch + 2,
+        0,
+        steps.splice(
+          steps.findIndex(step => 'limit' in step),
+          1,
+        )[0],
+      );
+    }
```

Comment:
We can leave it for now, but at least add a `TODO` here - this code is now barely readable.

Reply:
As this code is repeated three times here what do you think about exporting it to a function like for example
```ts
const spliceSteps = (indexOffset: number, stepName: 'skip' | 'limit') => {
      const [element] = steps.splice(
          steps.findIndex(step => stepName in step),
          1,
        )
      steps.splice(indexMatch + indexOffset, 0, element);
}
```

Reply:
Yeah, something like that. My proposition:
```ts
const wrapAtPhase = (index: number, stepName: 'limit' | 'skip') => {
  const pivot = steps.findIndex(step => stepName in step);
  steps.splice(index, 0, steps.splice(pivot, 1)[0]);
}
```

Reply:
Looks good! Changed.